services/personal/api.go

```diff
@@ -22,37 +23,44 @@ var (
 	ErrSignInvalidNumberOfParameters = errors.New("invalid number of parameters for personal_sign (2 or 3 expected)")
 )
 
-type metadata struct {
-	Data    interface{} `json:"data"`
-	Address string      `json:"account"`
+// SignParams required to sign messages
+type SignParams struct {
+	Data     interface{} `json:"data"`
+	Address  string      `json:"account"`
+	Password string      `json:"password"`
 }
 
-func newMetadata(rpcParams []interface{}) (*metadata, error) {
-	// personal_sign can be called with the following parameters
-	// 1) data to sign
-	// 2) account
-	// 3) (optional) password
-	// here, we always ignore (3) because we send a confirmation for the password to UI
-	if len(rpcParams) < 2 || len(rpcParams) > 3 {
-		return nil, ErrSignInvalidNumberOfParameters
-	}
-	data := rpcParams[0]
-	address := rpcParams[1].(string)
+// RecoverParams are for calling `personal_ecRecover`
+type RecoverParams struct {
+	Message   string
+	Signature string
+}
 
-	return &metadata{data, address}, nil
+// UnmarshalSignRPCParams puts the RPC params for `personal_sign` or `web3.personal.sign`
+// into SignParams
+func UnmarshalSignRPCParams(rpcParamsJSON string) (SignParams, error) {
```

Comment:
It looks like the better approach would be implementing [json.Unmarshaler](https://golang.org/pkg/encoding/json/#Unmarshaler) interface for these types (`SignParams` and `RecoverParams`). So the simple `json.Unmarshal()` will do the job.